File: hphe.m
%% ========================================================
%  Heat Pump & Heat Engine Modelling
%  Seville, Spain
%  Inputs: ST hourly output from ST module
%% ========================================================

clc; clear; close all;

%% ── 1. PARAMETERS 系统参数 ───────────────────────────────

% Temperature settings 温度设定
T_hot_HP  = 157 + 273.15;   % Heat pump output temp [K] 热泵出口温度
T_cold_HP = 70  + 273.15;   % Heat pump input (waste heat) [K] 废热温度
T_hot_HE  = 157 + 273.15;   % Heat engine hot side [K] 热机高温侧
T_cold_HE = 30  + 273.15;   % Heat engine cold side (ambient) [K] 冷侧

% Carnot efficiency correction factor
% (real systems achieve ~40-50% of Carnot)
epsilon_HP = 0.5;           % Heat pump Carnot correction 热泵修正系数
epsilon_HE = 0.2;           % Heat engine Carnot correction 热机修正系数

% Industrial demand 工业热需求
P_demand   = 10e6;           % need ST peak thermal power
t_op       = 10;             % Operating hours per day [h]

% Waste heat availability 废热参数
Q_waste_max = 5e6;           % Max available waste heat [W] = 5 MW
                             % (adjust based on project spec)

%% ── 2. CALCULATE COP AND EFFICIENCY ─────────────────────

% Heat Pump COP
COP_carnot = T_hot_HP / (T_hot_HP - T_cold_HP);
COP_real   = epsilon_HP * COP_carnot;

% Heat Engine efficiency
eta_carnot = 1 - T_cold_HE / T_hot_HE;
eta_real   = epsilon_HE * eta_carnot;

fprintf('=== Theoretical Performance ===\n');
fprintf('HP Carnot COP:       %.2f\n', COP_carnot);
fprintf('HP Real COP:         %.2f\n', COP_real);
fprintf('Engine Carnot eta:   %.2f%%\n', eta_carnot*100);
fprintf('Engine Real eta:     %.2f%%\n', eta_real*100);
fprintf('================================\n\n');

%% ── 3. LOAD ST OUTPUT 读取ST模块输出 ─────────────────────
%
% Q_ST_hourly should come from your ST simulation (8760 values)
% If not available yet, generate a simplified estimate:

% Option A: Load from ST module output
% load('ST_output.mat', 'Q_hourly');
% Q_ST_hourly = Q_hourly;   % [Wh per hour]

% Option B: Simplified sinusoidal estimate (placeholder)
% Replace this with real ST output when available
t_vec = (1:8760)';
Q_ST_peak = 10.1e6;         % Peak ST output from ST sizing [W]
Q_ST_hourly = zeros(8760,1);
for i = 1:8760
    hour_of_day = mod(i-1, 24);
    day_of_year = floor((i-1)/24) + 1;
    % Solar availability: 7am-6pm, seasonal variation
    if hour_of_day >= 7 && hour_of_day <= 18
        solar_hour  = sin(pi*(hour_of_day-7)/11);
        seasonal    = 0.7 + 0.3*sin(2*pi*(day_of_year-80)/365);
        Q_ST_hourly(i) = Q_ST_peak * solar_hour * seasonal;
    end
end
% Convert to Wh (already per-hour values)

% Build hourly demand profile (only during operating hours)
Q_demand_hourly = zeros(8760, 1);
for i = 1:8760
    hour_of_day = mod(i-1, 24);
    % Operating schedule: 8am-6pm (adjust to your project)
    if hour_of_day >= 8 && hour_of_day <= 18
        Q_demand_hourly(i) = P_demand;   % [W]
    end
end

%% ── 4. HOURLY SIMULATION 逐时仿真 ────────────────────────

% Output arrays
Q_HP_hourly     = zeros(8760, 1);   % Heat pump heat output [Wh]
W_elec_hourly   = zeros(8760, 1);   % Heat pump electricity input [Wh]
Q_waste_hourly  = zeros(8760, 1);   % Waste heat consumed by HP [Wh]
W_engine_hourly = zeros(8760, 1);   % Heat engine electricity output [Wh]
Q_excess_hourly = zeros(8760, 1);   % Excess ST heat to engine [Wh]
Q_reject_hourly = zeros(8760, 1);   % Engine heat rejection [Wh]

for t = 1:8760

    Q_ST  = Q_ST_hourly(t);       % ST output this hour [Wh]
    Q_dem = Q_demand_hourly(t);   % Demand this hour [Wh]

    if Q_dem == 0
        % No demand (outside operating hours)
        continue
    end

    if Q_ST < Q_dem
        %% ── HEAT PUMP MODE 热泵模式 ──────────────────────
        % ST not enough → heat pump covers the gap

        Q_gap = Q_dem - Q_ST;           % Heat gap to fill [Wh]

        % Check waste heat availability
        Q_waste_avail = min(Q_waste_max, Q_gap * (1 - 1/COP_real));

        % Heat pump output
        Q_HP = COP_real * (Q_gap / COP_real) * COP_real;
        % Simplified: HP needs to produce Q_gap
        Q_HP = Q_gap;

        % Electricity consumed by heat pump
        W_elec = Q_HP / COP_real;

        % Waste heat consumed
        Q_waste_used = Q_HP - W_elec;

        % Store results
        Q_HP_hourly(t)    = Q_HP;
        W_elec_hourly(t)  = W_elec;
        Q_waste_hourly(t) = Q_waste_used;

    else
        %% ── HEAT ENGINE MODE 热机模式 ────────────────────
        % ST excess → heat engine generates electricity

        Q_excess = Q_ST - Q_dem;        % Excess heat [Wh]

        % Engine electricity output
        W_out    = eta_real * Q_excess;

        % Heat rejected to cold side
        Q_reject = Q_excess - W_out;

        % Store results
        Q_excess_hourly(t) = Q_excess;
        W_engine_hourly(t) = W_out;
        Q_reject_hourly(t) = Q_reject;
    end
end

%% ── 5. ANNUAL RESULTS 年度结果 ───────────────────────────

% Heat Pump results
E_HP_heat_MWh  = sum(Q_HP_hourly)    / 1e6;   % Annual heat from HP [MWh]
E_HP_elec_MWh  = sum(W_elec_hourly)  / 1e6;   % Annual electricity to HP [MWh]
E_waste_MWh    = sum(Q_waste_hourly) / 1e6;   % Annual waste heat used [MWh]

% Heat Engine results
E_engine_MWh   = sum(W_engine_hourly) / 1e6;  % Annual electricity generated [MWh]
E_excess_MWh   = sum(Q_excess_hourly) / 1e6;  % Annual excess ST heat [MWh]
E_reject_MWh   = sum(Q_reject_hourly) / 1e6;  % Annual heat rejected [MWh]

% Hours of operation
HP_hours     = sum(Q_HP_hourly > 0);
engine_hours = sum(W_engine_hourly > 0);

fprintf('=== Annual Results ===\n\n');
fprintf('-- Heat Pump --\n');
fprintf('Heat output:          %.1f MWh/year\n', E_HP_heat_MWh);
fprintf('Electricity input:    %.1f MWh/year\n', E_HP_elec_MWh);
fprintf('Waste heat used:      %.1f MWh/year\n', E_waste_MWh);
fprintf('Operating hours:      %d h/year\n',     HP_hours);
fprintf('Effective COP:        %.2f\n\n',         E_HP_heat_MWh/E_HP_elec_MWh);

fprintf('-- Heat Engine --\n');
fprintf('Electricity generated: %.1f MWh/year\n', E_engine_MWh);
fprintf('Excess heat input:     %.1f MWh/year\n', E_excess_MWh);
fprintf('Heat rejected:         %.1f MWh/year\n', E_reject_MWh);
fprintf('Operating hours:       %d h/year\n',     engine_hours);
fprintf('======================\n');

%% ── 6. MONTHLY BREAKDOWN 逐月分布 ────────────────────────

days_per_month  = [31,28,31,30,31,30,31,31,30,31,30,31];
hours_per_month = days_per_month * 24;
E_HP_monthly     = zeros(12,1);
E_engine_monthly = zeros(12,1);
h = 1;
for m = 1:12
    idx = h : h + hours_per_month(m) - 1;
    E_HP_monthly(m)     = sum(Q_HP_hourly(idx))     / 1e6;
    E_engine_monthly(m) = sum(W_engine_hourly(idx)) / 1e6;
    h = h + hours_per_month(m);
end

%% ── 7. PLOTS 画图 ────────────────────────────────────────

months = {'Jan','Feb','Mar','Apr','May','Jun',...
          'Jul','Aug','Sep','Oct','Nov','Dec'};

% Plot 1: Monthly HP heat output vs Engine electricity output
figure('Name','Monthly HP and Engine Output');
bar(1:12, [E_HP_monthly, E_engine_monthly]);
legend('HP Heat Output (MWh)', 'Engine Elec Output (MWh)');
xlabel('Month'); ylabel('Energy (MWh)');
title('Monthly Heat Pump & Engine Output - Seville');
xticklabels(months); grid on;

% Plot 2: COP sensitivity to temperature
figure('Name','COP vs Temperature');
T_cold_range = 50:5:100;
COP_curve    = epsilon_HP * T_hot_HP ./ (T_hot_HP - (T_cold_range+273.15));
plot(T_cold_range, COP_curve, 'b-o', 'LineWidth', 1.5);
xlabel('Waste Heat Temperature T_{cold} (°C)');
ylabel('Real COP');
title('Heat Pump COP vs Waste Heat Temperature');
grid on;

% Plot 3: Sample week hourly operation
figure('Name','Sample Week Operation');
t_sample = 1:168;   % first week
subplot(2,1,1);
area(t_sample, Q_ST_hourly(t_sample)/1e3, 'FaceColor',[1 0.8 0.2]);
hold on;
area(t_sample, Q_HP_hourly(t_sample)/1e3, 'FaceColor',[0.2 0.6 1]);
plot(t_sample, Q_demand_hourly(t_sample)/1e3, 'r--','LineWidth',1.5);
legend('ST Output (kWh)','HP Output (kWh)','Demand (kWh)');
xlabel('Hour'); ylabel('Energy (kWh)');
title('Heat Supply vs Demand - First Week');
grid on;

subplot(2,1,2);
bar(t_sample, W_engine_hourly(t_sample)/1e3, 'FaceColor',[0.2 0.8 0.4]);
xlabel('Hour'); ylabel('Electricity (kWh)');
title('Engine Electricity Output - First Week');
grid on;

File: hphe.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec


def simulate_heat_pump_engine(
    Q_ST_hourly: np.ndarray,
    Q_demand_hourly:  np.full(8760, 10e6),
    T_cold_HP: float = 70.0,
    T_hot_HP: float = 157.0,
    T_cold_HE: float = 40.0,
    T_hot_HE: float = 157.0,
    COP_real: float = 2.32,
    epsilon_HE: float = 0.20,
    Q_waste_max: float = 5e6,
    plot: bool = True,
    save_fig: bool = True,
    fig_path: str = 'hp_engine_results.png'
) -> dict:
    """
    Simulate hourly operation of High-Temperature Heat Pump (HTHP)
    and Stirling Heat Engine in a hybrid solar industrial heat system.
    Optionally generates and saves result plots.

    Heat Pump parameters based on:
    Enerin HoegTemp, IEA HPT Annex 58, Table 1
    (Tsource=80->50C, Tsink=154->160C, COP=2.2-2.45)

    Parameters
    ----------
    Q_ST_hourly     : np.ndarray  Hourly ST heat output [Wh], shape (8760,)
    Q_demand_hourly : np.ndarray  Hourly industrial heat demand [Wh], shape (8760,)
    T_cold_HP       : float       Heat pump cold side temperature [C], default 70
    T_hot_HP        : float       Heat pump hot side temperature [C], default 157
    T_cold_HE       : float       Heat engine cold side temperature [C], default 40
    T_hot_HE        : float       Heat engine hot side temperature [C], default 157
    COP_real        : float       Actual heat pump COP [-], default 2.32
    epsilon_HE      : float       Heat engine Carnot correction factor [-], default 0.20
    Q_waste_max     : float       Max available waste heat [W], default 5e6
    plot            : bool        Whether to generate plots, default True
    save_fig        : bool        Whether to save figure to file, default True
    fig_path        : str         File path for saved figure, default 'hp_engine_results.png'

    Returns
    -------
    dict with keys:
        'Q_HP_hourly'     : np.ndarray  Heat pump heat output [Wh]
        'W_elec_hourly'   : np.ndarray  Heat pump electricity input [Wh]
        'Q_waste_hourly'  : np.ndarray  Waste heat consumed [Wh]
        'W_engine_hourly' : np.ndarray  Engine electricity output [Wh]
        'Q_excess_hourly' : np.ndarray  Excess ST heat to engine [Wh]
        'Q_reject_hourly' : np.ndarray  Engine heat rejection [Wh]
        'annual'          : dict        Annual summary statistics
        'monthly'         : dict        Monthly breakdown
        'performance'     : dict        COP and efficiency values
    """

    # ── 1. Temperature conversion ─────────────────────────────────
    T_cold_HP_K = T_cold_HP + 273.15
    T_hot_HP_K  = T_hot_HP  + 273.15
    T_cold_HE_K = T_cold_HE + 273.15
    T_hot_HE_K  = T_hot_HE  + 273.15

    # ── 2. Derived performance parameters ─────────────────────────
    COP_carnot = T_hot_HP_K / (T_hot_HP_K - T_cold_HP_K)
    epsilon_HP = COP_real / COP_carnot
    eta_carnot = 1.0 - T_cold_HE_K / T_hot_HE_K
    eta_real   = epsilon_HE * eta_carnot

    # ── 3. Hourly simulation arrays ───────────────────────────────
    Q_HP_hourly     = np.zeros(8760)
    W_elec_hourly   = np.zeros(8760)
    Q_waste_hourly  = np.zeros(8760)
    W_engine_hourly = np.zeros(8760)
    Q_excess_hourly = np.zeros(8760)
    Q_reject_hourly = np.zeros(8760)

    # ── 4. Main hourly loop ───────────────────────────────────────
    for t in range(8760):
        Q_ST  = Q_ST_hourly[t]
        Q_dem = Q_demand_hourly[t]

        if Q_dem == 0:
            continue

        if Q_ST < Q_dem:
            # Heat Pump mode: ST insufficient, HP covers the gap
            Q_HP              = Q_dem - Q_ST
            W_elec            = Q_HP / COP_real
            Q_waste_used      = min(Q_HP - W_elec, Q_waste_max)
            Q_HP_hourly[t]    = Q_HP
            W_elec_hourly[t]  = W_elec
            Q_waste_hourly[t] = Q_waste_used
        else:
            # Heat Engine mode: ST excess drives engine
            Q_excess              = Q_ST - Q_dem
            W_out                 = eta_real * Q_excess
            Q_reject              = Q_excess - W_out
            Q_excess_hourly[t]    = Q_excess
            W_engine_hourly[t]    = W_out
            Q_reject_hourly[t]    = Q_reject

    # ── 5. Annual summary ─────────────────────────────────────────
    annual = {
        'HP_heat_MWh'    : float(np.sum(Q_HP_hourly)     / 1e6),
        'HP_elec_MWh'    : float(np.sum(W_elec_hourly)   / 1e6),
        'waste_heat_MWh' : float(np.sum(Q_waste_hourly)  / 1e6),
        'engine_elec_MWh': float(np.sum(W_engine_hourly) / 1e6),
        'excess_heat_MWh': float(np.sum(Q_excess_hourly) / 1e6),
        'heat_reject_MWh': float(np.sum(Q_reject_hourly) / 1e6),
        'HP_hours'       : int(np.sum(Q_HP_hourly > 0)),
        'engine_hours'   : int(np.sum(W_engine_hourly > 0)),
    }
    annual['COP_effective'] = (
        annual['HP_heat_MWh'] / annual['HP_elec_MWh']
        if annual['HP_elec_MWh'] > 0 else 0.0
    )

    # ── 6. Monthly breakdown ──────────────────────────────────────
    days_per_month  = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    hours_per_month = [d * 24 for d in days_per_month]
    HP_monthly      = np.zeros(12)
    engine_monthly  = np.zeros(12)
    h = 0
    for m in range(12):
        idx             = slice(h, h + hours_per_month[m])
        HP_monthly[m]     = np.sum(Q_HP_hourly[idx])     / 1e6
        engine_monthly[m] = np.sum(W_engine_hourly[idx]) / 1e6
        h += hours_per_month[m]

    monthly = {
        'HP_heat_MWh'    : HP_monthly,
        'engine_elec_MWh': engine_monthly,
    }

    # ── 7. Performance parameters ─────────────────────────────────
    performance = {
        'COP_carnot': round(COP_carnot, 3),
        'COP_real'  : round(COP_real,   3),
        'epsilon_HP': round(epsilon_HP, 3),
        'eta_carnot': round(eta_carnot, 3),
        'eta_real'  : round(eta_real,   3),
        'epsilon_HE': round(epsilon_HE, 3),
    }

    # ── 8. Plotting ───────────────────────────────────────────────
    if plot:
        months_label = ['Jan','Feb','Mar','Apr','May','Jun',
                        'Jul','Aug','Sep','Oct','Nov','Dec']

        fig = plt.figure(figsize=(16, 18))
        fig.suptitle(
            'Heat Pump & Stirling Engine — Annual Simulation Results\n'
            'Seville | 10 MW Industrial Load | Enerin HoegTemp',
            fontsize=14, fontweight='bold', y=0.98
        )
        gs = gridspec.GridSpec(3, 2, figure=fig, hspace=0.45, wspace=0.35)

        # Plot 1: Monthly bar chart
        ax1 = fig.add_subplot(gs[0, 0])
        x, w = np.arange(12), 0.35
        ax1.bar(x - w/2, monthly['HP_heat_MWh'],     w,
                label='HP Heat Output',     color='#E07B39')
        ax1.bar(x + w/2, monthly['engine_elec_MWh'], w,
                label='Engine Elec Output', color='#3A7EBF')
        ax1.set_xticks(x)
        ax1.set_xticklabels(months_label)
        ax1.set_xlabel('Month')
        ax1.set_ylabel('Energy (MWh)')
        ax1.set_title('Monthly HP Heat & Engine Electricity Output')
        ax1.legend()
        ax1.grid(axis='y', alpha=0.4)

        # Plot 2: Pie chart
        ax2    = fig.add_subplot(gs[0, 1])
        vals   = [annual['HP_heat_MWh'],    annual['engine_elec_MWh'],
                  annual['waste_heat_MWh'], annual['heat_reject_MWh']]
        lbls   = ['HP Heat\nOutput', 'Engine Elec\nOutput',
                  'Waste Heat\nUsed', 'Engine Heat\nRejected']
        colors = ['#E07B39', '#3A7EBF', '#5BAD6F', '#C0392B']
        non_zero = [(v, l, c) for v, l, c in zip(vals, lbls, colors) if v > 0]
        if non_zero:
            v_p, l_p, c_p = zip(*non_zero)
            ax2.pie(v_p, labels=l_p, colors=c_p,
                    autopct='%1.1f%%', startangle=140,
                    textprops={'fontsize': 9})
        ax2.set_title('Annual Energy Distribution')

        # Plot 3: First week stack
        ax3     = fig.add_subplot(gs[1, :])
        t_week  = np.arange(168)
        Q_ST_kW = Q_ST_hourly[:168]        / 1e3
        Q_HP_kW = Q_HP_hourly[:168]        / 1e3
        Q_dm_kW = Q_demand_hourly[:168]    / 1e3
        Q_en_kW = W_engine_hourly[:168]    / 1e3
        ax3.fill_between(t_week, 0, Q_ST_kW,
                         alpha=0.6, color='#F4D03F', label='ST Output (kWh)')
        ax3.fill_between(t_week, Q_ST_kW, Q_ST_kW + Q_HP_kW,
                         alpha=0.6, color='#E07B39', label='HP Output (kWh)')
        ax3.plot(t_week, Q_dm_kW,
                 'r--', linewidth=1.8, label='Demand (kWh)')
        ax3.fill_between(t_week, 0, -Q_en_kW,
                         alpha=0.5, color='#3A7EBF', label='Engine Elec Out (kWh)')
        ax3.axhline(0, color='black', linewidth=0.8)
        ax3.set_xlabel('Hour (first 7 days)')
        ax3.set_ylabel('Power (kW)')
        ax3.set_title('Heat Supply Stack vs Demand — First Week')
        ax3.legend(loc='upper right')
        ax3.grid(alpha=0.3)

        # Plot 4: COP sensitivity
        ax4        = fig.add_subplot(gs[2, 0])
        T_cold_rng = np.arange(30, 120, 5)
        COP_curve  = np.clip(
            performance['epsilon_HP'] * T_hot_HP_K /
            (T_hot_HP_K - (T_cold_rng + 273.15)),
            0, 10
        )
        ax4.plot(T_cold_rng, COP_curve, 'b-o', markersize=4, linewidth=1.5)
        ax4.axvline(x=70,  color='red',    linestyle='--', linewidth=1.2,
                    label='Project T_cold = 70°C')
        ax4.axhline(y=COP_real, color='orange', linestyle='--', linewidth=1.2,
                    label=f'COP = {COP_real:.2f}')
        ax4.set_xlabel('Waste Heat Temperature T_cold (°C)')
        ax4.set_ylabel('Real COP [-]')
        ax4.set_title('HP COP vs Waste Heat Temperature')
        ax4.legend(fontsize=8)
        ax4.grid(alpha=0.4)

        # Plot 5: Annual summary bar
        ax5      = fig.add_subplot(gs[2, 1])
        cat      = ['HP Heat\nOutput', 'HP Elec\nInput', 'Waste Heat\nUsed',
                    'Engine Elec\nOutput', 'Excess ST\nHeat', 'Heat\nRejected']
        vals_bar = [annual['HP_heat_MWh'],    annual['HP_elec_MWh'],
                    annual['waste_heat_MWh'], annual['engine_elec_MWh'],
                    annual['excess_heat_MWh'],annual['heat_reject_MWh']]
        bar_colors = ['#E07B39','#E07B39','#5BAD6F',
                      '#3A7EBF','#F4D03F','#C0392B']
        bars = ax5.bar(cat, vals_bar, color=bar_colors,
                       alpha=0.85, edgecolor='white')
        for bar, val in zip(bars, vals_bar):
            if val > 0:
                ax5.text(bar.get_x() + bar.get_width() / 2,
                         bar.get_height() + max(vals_bar) * 0.01,
                         f'{val:.0f}', ha='center', va='bottom', fontsize=8)
        ax5.set_ylabel('Energy (MWh/year)')
        ax5.set_title('Annual Energy Summary')
        ax5.tick_params(axis='x', labelsize=8)
        ax5.grid(axis='y', alpha=0.4)

        # KPI text box
        kpi_text = (
            f"Key Performance Indicators\n"
            f"{'─'*32}\n"
            f"COP Carnot:      {performance['COP_carnot']:.2f}\n"
            f"COP Real:        {performance['COP_real']:.2f}  "
            f"(epsilon = {performance['epsilon_HP']:.2f})\n"
            f"COP Effective:   {annual['COP_effective']:.2f}\n"
            f"eta Carnot (HE): {performance['eta_carnot']*100:.1f}%\n"
            f"eta Real (HE):   {performance['eta_real']*100:.1f}%\n"
            f"HP Hours/year:   {annual['HP_hours']} h\n"
            f"Engine Hours/yr: {annual['engine_hours']} h"
        )
        fig.text(0.5, 0.005, kpi_text, ha='center', va='bottom',
                 fontsize=9, family='monospace',
                 bbox=dict(boxstyle='round', facecolor='#EBF3FB',
                           edgecolor='#2E74B5', alpha=0.9))

        if save_fig:
            plt.savefig(fig_path, dpi=150,
                        bbox_inches='tight', facecolor='white')
            print(f"Figure saved as {fig_path}")

        plt.show()

    # ── 9. Return ─────────────────────────────────────────────────
    return {
        'Q_HP_hourly'    : Q_HP_hourly,
        'W_elec_hourly'  : W_elec_hourly,
        'Q_waste_hourly' : Q_waste_hourly,
        'W_engine_hourly': W_engine_hourly,
        'Q_excess_hourly': Q_excess_hourly,
        'Q_reject_hourly': Q_reject_hourly,
        'annual'         : annual,
        'monthly'        : monthly,
        'performance'    : performance,
    }
